### ISSUE TITLE
"Publish on next push"

### ISSUE THREAD

**Issue by fahhem** [Reviewable/Reviewable#1104]
When an author is replying to a review, often it's a series of "Done" type messages. However, those messages aren't accurate until the next push, so an author has to decide to either:

* Git push and then go back to the UI and hit Publish
* Publish, then git push (and hope nothing stops/delays that)

It would be nice to be able to, if I'm the author, get an option to "Publish on next Push". Then I can go and deal with git pushing, then move on to the next task rather than having to go back to the UI.

**Comment by jwnimmer-tri** [comment]
One way to weave this in might be one more disposition in the disposition drop-down, instead of a separate setting.  Something like "Fix Pending" which would behave like "Pondering, automatically published as Satisfied upon next push".  (The "Done" button would switch to that disposition by default.)

**Comment by pkaminski** [comment]
Ooh, I like it!  Perhaps we could even replace Pondering with Fix Pending for PR authors so we don't bloat the disposition menu too much?  Is there much need to use Pondering for authors?

**Comment by ajwerner** [comment]
I think that disposition seems like a bad fit here. I use the `Done` button a lot when responding to feedback. It'd be nice to have that result in comments which are in that disposition or maybe more generally to toggle whether the default disposition is this "Fix Pending" vs the current default ("Discussing"?).  Disposition is useful, and if this thing is part of disposition it's not clear how it'd interact with the disposition I'd like the comment to be in upon being posted.

One approach might be to make that the default for `Done`, and I think that'd be pretty good, but I think it might be very surprising if only the things that I have actually fixed get published when I push vs comments that aren't resolved. I very well may have things to say in the reply to the review, both at the top level and on still open threads. That reply is going to reference code that I'm about to push. If I have to go back to push publish for some of those comments, then I'm not sure this is providing much value.

**Comment by pkaminski** [comment]
Yeah, that's fair -- unless we want all on-push comments to be sent with a Satisfied disposition we can't really do it this way.  And while you could always publish your non-on-push varied-disposition drafts first, then push to have the rest published, that's awkward for both sender and receiver.  Most often, you'll be sending some "done" and some "no, let's not" replies in a round, together with your pushed code.

Which brings us back to a separate per-draft toggle, which would work but I'm not convinced is worth the complication for what is a bit of a corner use case (ad-hoc send of a subset of comments without having to click them one-by-one).  Perhaps that would be better addressed by, e.g., letting you long-click the ad-hoc send button to add a draft to a send-now set, and turning the publish button into a "Publish Marked" variant.

Which then brings us back to the original proposal, a toggle to "Publish on Push".  I think we'd want to restrict it to only the PR author, both because typically they'll be the only ones pushing and because there'll be no space in the button for the approve/block suffix.  (If a reviewer wants to push and publish they can do it the old manual way.)  This also happens to be the (likely) easiest to implement, so +1 for that.

Any further feedback, anyone?

**Comment by pkaminski** [comment]
Fleshing out a potential design:

We'll add a "Wait until your next push to this branch to publish" checkbox to the Publish options dropdown.  When checked, the button turns into "Publish on push".  The state is sticky per-user per-review, and the last click sets the default for new reviews.  The checkbox is enabled only if the user is the PR author and the repo is connected; otherwise, it's disabled and unchecked with an explanation in fine print as to why.

Clicking "Publish on push" does the usual prep work then queues a bundle with the server.  The publish button changes to a disabled button (or info label?) that says "Waiting for push to publish".  The next push to the PR's head branch by this user will automatically publish the bundle exactly as if the user had done it manually in the client, and return the user's review to its normal post-publish state.  (Note: the author and committer of the commits doesn't matter, only the pusher.)  If the user tries to edit a draft or create a new one, or basically does anything that would change their personal review state, the pending publish is automatically and silently canceled and the review goes back to its normal state.

If a publish bundle is queued for more than 30 minutes without a push to release it, we cancel the publish, return the review to its normal state with all drafts intact, and send an email notification to the user advising them that this happened.  Similarly, if the repo is disconnected (whether intentionally or due to an error), all pending publications are canceled and notifications sent out.

**Comment by fahhem** [comment]
I suggest that the "Publish on push", after briefly being disabled while we execute the action, should turn into a (yellow?) "Cancel publish". Otherwise users unfamiliar with the automatic-cancellation may panic when they hit publish, then realize there's a mistake in a comment they think is sort-of-published.

**Comment by earlAchromatic** [comment]
One idea that would let you achieve 'publish on push' but with any resulting disposition is to leverage the magic keywords, not the disposition dropdown, and use a new syntax like: `on-push:satisfied`, `on-push:discussing` or whatever disposition you want. Other magic keywords should be valid too like: `on-push:done`.

You publish like normal, but comments matching this pattern don't actually show up until the push happens as described above. You would apply the 'Waiting for push to publish' message and timeouts to these as well.

**Comment by earlAchromatic** [comment]
I like the per-draft-toggle idea and I can imagine a minimal UI implementation for doing published subsets using it. We add a small, unobtrusive checkbox in the discussion header or ribbon label. When any one comment checkbox in the review is activated, publishing immediately goes into "subset" mode where only other comments with the checkbox ticked would be sent by hitting the general publish button. The publish preview would be populated with only the selected subset of comments.

![image](https://github.com/Reviewable/Reviewable/assets/68669571/ce7c5208-9461-4be1-a2e4-0b99356032ac)

We add an indicator on the publish button that you are in subset mode, and a utility for selecting all comments so you can deselect only the one or two that you don't want published, or revert to regular publishing mode by deselecting all comments. This ‘to publish’ selection workflow will be greatly improved by having these checkboxes also show up in the discussion list in the coming sidebar (far less bouncing around the review).

This design would eliminate the need for the pondering disposition and the ad-hoc publish button. This approach to managing drafts would be superior to our current design as it would let you use the same workflow just with greater flexibility for what gets published at what time.

If we like the minimal UI concept, we could also put both publish subsetting and publish on push (as another small checkbox input in a different color) right next to it in the ribbon label or header, and ditch the magic keyword idea I wrote above.

**Comment by jwnimmer-tri** [comment]
> Fleshing out a potential design:
>
> We'll add a "Wait until your next push to this branch to publish" checkbox to the Publish options dropdown...

I like this simple design a lot.  It solves the problem of deferring comment posting to align with the the git push, so the reviewer will be presented with a consistent set of information all at once.

After some thought, my point of view now is that any _per-discussion_ toggle for "publish on push" will be too difficult to teach to users, be it with disposition, keywords, or UI elements.  The review-wide toggle is the right layer to attack the problem, when our goal is to sync with the "git push".

For advanced users who want to withhold a specific comment from publishing _at all_, even on a push, that's what "Pondering" is already there for, and (I assume) will continue to work just fine.

The only problem that "publish on push" doesn't solve is "please publish these 5 comments as a single message right now, and keep the remaining 25 comments open as drafts for me to circle back to later, myself".  We would be left with just the "publish 1 comment" button that already exists (in the lower right corner of the discussion box), which we'd need to click 5 times and send 5 emails.  A possible improvement there is to have that "publish one comment" button do some kind of batching, to reduce the number of emails.  That's a different story than this ticket, though.  That story is "publish a subset of my comments immediately, with a single posting", and would make sense even without any "git push" coordination.  This story is "publish the remainder of my comments upon the next git push".

**Comment by fahhem** [comment]
I don't think we need to worry much about a subset/partial publish feature in combination with "Publish on Push". This is because, if you're doing a "Publish on Push", then you're going to push soon (within a half hour) or we'll cancel the whole thing. The reason for cancelling will likely be that you failed to push for one reason or another (distraction, precommit checks, etc) and you'll have this PR as awaiting your action on your dashboard anyway, and you'll come back.

There could be an argument for "Publish on Push" having an option to publish on the next push or in 30 minutes, whichever comes first (like a "Publish soon" or "Publish soon for sure"), but I don't know if that option would be useful and as self-explanatory as a simple "Publish > On Push"

**Comment by pkaminski** [comment]
> A possible improvement there is to have that "publish one comment" button do some kind of batching, to reduce the number of emails. That's a different story than this ticket, though.

It's not necessarily a _wholly_ different story, in that (I think) the most common reason for publishing some messages separately is that you want to send some questions or pushback without sending all your replies that will only make sense once you push a code update.  However, I believe that "publish on push" is core functionality that needs to be easily discoverable and accessible, and so far every proposed approach to combining it with the early-batch-send feature has made it unacceptably more complicated to use.  That's not to say that there isn't some design that could hit the sweet spot, though.

> This design [per-draft toggle and publish subset] would eliminate the need for the pondering disposition and the ad-hoc publish button.

Not entirely.  I think you're right that the pondering disposition could go (though we'd probably want to keep the shortcut), as long as we allow combining "publish subset" with "publish on push" and "publish & agree/block".  (Perhaps to save space we can just say something like "publish M/N" on the button.)  The ad-hoc publish button we can't get rid of quite yet, though, as it's used to send comments with no Reviewable wrappers (review state, etc.), which is required by some GitHub bots.  However this is hard to discover, so if we had a solution for https://github.com/Reviewable/Reviewable/issues/952 I'd be glad to get rid of the ad-hoc single-comment publish.

> publish on push (as another small checkbox input in a different color) right next to it in the ribbon label

While pleasingly orthogonal and consistent, I think I'm with Jeremy that this just feels too fiddly.  It would also trivially lead to a situation where you click publish, some messages go out in a batch, then another batch goes out when you push.  That seems like it would be an easy mistake to make by forgetting to tick an "on push" checkbox and almost certainly not what was intended.

> There could be an argument for "Publish on Push" having an option to publish on the next push or in 30 minutes, whichever comes first (like a "Publish soon" or "Publish soon for sure"), but I don't know if that option would be useful and as self-explanatory as a simple "Publish > On Push"

We could add that later (as replacement or sub-option) if we hear demand, but I wouldn't lead with it -- as you say, it's not as clear-cut.

At this point I'm inclined to run with the simple "publish on push" design above and punt subsetting to a separate issue to explore potential designs.

**Comment by fahhem-jurin** [comment]
🎉🎉🎉🎊🎊🎊🙌🙌🙌🎊🎊🎊🎉🎉🎉

**Comment by pkaminski** [comment]
😁 We'll likely get it deployed early next week as part of a bigger release.
